### PR TITLE
feat(log-collector): add watch-based pod discovery with polling fallback

### PR DESCRIPTION
**Pull request by ygrishajev**
## Why

The log-collector uses polling to detect pod additions and removals, which adds latency and unnecessary API load. The K8s Watch API provides real-time events but may not be available in all RBAC configurations.

## What

Adds K8s Watch-based pod discovery as the primary mechanism, with automatic fallback to polling when watch is unavailable.

**Watch behavior:**
- Watch works → use it for real-time ADDED/MODIFIED/DELETED events, no polling
- Watch fails 403 → fall back to polling permanently (RBAC missing `watch` verb)
- Watch fails other → poll for 30s, retry watch, repeat indefinitely
- Watch ends normally (server timeout) → reconnect immediately

**Implementation:**
- Uses AsyncGenerators to compose the watch/poll/retry flow declaratively
- `podEventStream()` orchestrates: try watch → fallback → retry — visible in ~15 lines
- `AsyncChannel` bridges K8s Watch push-based callbacks to pull-based async iteration (generic utility in `src/lib/`)
- DELETED events bypass readiness checks (deleted pods aren't ready)

**Infrastructure:**
- Register `Watch` client in `k8s-client.provider.ts`
- Add `watch` verb to RBAC `role.yaml`

**Testing:**
- Manually verified all flows on local K8s cluster:
  - Flow 1: Watch detects pod additions (ADDED) and removals (DELETED) in real-time
  - Flow 3: 403 triggers permanent polling with `POD_WATCH_FORBIDDEN` warning
- Unit tests: 66 passing across 8 suites (including 6 AsyncChannel tests)

## Summary by CodeRabbit

* **New Features**
  * Pod discovery now uses an async event stream driven by Kubernetes Watch with automatic polling fallback and improved deduplication/tracking.
  * Added an AsyncChannel utility for async producer-consumer event streaming.

* **Chores**
  * Separated pod and pod-log RBAC permissions.
  * Improved startup diagnostic logging.

* **Tests**
  * Expanded pod discovery tests and added AsyncChannel tests covering watch/polling behaviors and edge cases.

* **Documentation**
  * Updated README to document watch/poll fallback and a POD_POLL_INTERVAL_MS setting.

### CODE REVIEW

File: apps/log-collector/src/services/pod-discovery/pod-discovery.service.ts [hunk 1]
```diff
@@ -83,50 +105,182 @@ export class PodDiscoveryService {
...
+            }
+            return;
+          }
+
+          const podInfo = this.processPod(apiObj);
+          if (!podInfo) return;
+
+          if ((phase === "ADDED" || phase === "MODIFIED") && !this.controllers.has(podInfo.podName)) {
+            channel.push({ type: "added", podInfo });
+          }
```
[the remark this answers is not in the document]
Comment (ygrishajev) [reply apps/log-collector/src/services/pod-discovery/pod-discovery.service.ts:200]: @coderabbitai Won't fix — from a log collection perspective, untracking a not-ready pod would stop collecting logs while the container is still running (e.g., crash loop). The K8s log stream stays open regardless of readiness status. We want to keep collecting until the pod is actually DELETED or the stream closes naturally. The polling behavior (untracking on not-ready) is arguably the less correct one here.

File: apps/log-collector/src/lib/async-channel/async-channel.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+ * // Consumer side (async iteration)
+ * for await (const value of channel) {
+ *   console.log(value);
+ * }
+ * ```
+ */
+export class AsyncChannel<T> {
+  private queue: T[] = [];
+
+  private waiting: ((result: IteratorResult<T>) => void) | null = null;
```
[the remark this answers is not in the document]
Comment (ygrishajev) [reply apps/log-collector/src/lib/async-channel/async-channel.ts:27]: @coderabbitai Won't fix — `yield*` calls `next()` sequentially, always waiting for the previous result before requesting the next one. There are no concurrent consumers of the channel in our usage. The single-resolver design is intentional for simplicity.

File: apps/log-collector/k8s/role.yaml [hunk 3]
```diff
@@ -5,7 +5,10 @@ metadata:
   name: tenant-pod-reader
 rules:
   - apiGroups: [""]
-    resources: ["pods", "pods/log"]
+    resources: ["pods"]
+    verbs: ["get", "list", "watch"]
+  - apiGroups: [""]
+    resources: ["pods/log"]
     verbs: ["get", "list"]
```
[the remark this answers is not in the document]
Comment (ygrishajev) [reply apps/log-collector/k8s/role.yaml:12]: @coderabbitai Won't fix — this is a local development/testing setup, not a production RBAC configuration. The `default` service account usage is intentional for simplicity in the dev environment.

File: apps/log-collector/k8s/role.yaml [hunk 4]
```diff
@@ -6,7 +6,7 @@ metadata:
 rules:
   - apiGroups: [""]
     resources: ["pods", "pods/log"]
-    verbs: ["get", "list"]
+    verbs: ["get", "list", "watch"]
```
[the remark this answers is not in the document]
Comment (ygrishajev) [reply apps/log-collector/k8s/role.yaml:9]: @coderabbitai check if this is still valid

File: apps/log-collector/src/services/pod-discovery/pod-discovery.service.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ygrishajev) [reply apps/log-collector/src/services/pod-discovery/pod-discovery.service.ts:200]: @coderabbitai any concerns here?

File: apps/log-collector/src/services/pod-discovery/pod-discovery.service.ts [hunk 5]
```diff
@@ -57,116 +76,278 @@ export class PodDiscoveryService {
...
+      this.loggerService.info({ event: "POLLING_STOPPED" });
+    }
+
+    return lastResourceVersion;
+  }
+
+  /** Polls for a fixed duration, then returns the latest resourceVersion so the caller can resume watch. */
+  private async *timedPollEvents(durationMs: number): AsyncGenerator<PodEvent, string | undefined> {
+    const ac = new AbortController();
+    const timeout = setTimeout(() => ac.abort(), durationMs);
```
Comment (stalniy) [inline apps/log-collector/src/services/pod-discovery/pod-discovery.service.ts:271]: **nitpick:** there is `AbortSignal.timeout` helper for this https://developer.mozilla.org/en-US/docs/Web/API/AbortSignal/timeout_static

**Review by stalniy — APPROVED** [review]
